infor-design/enterprise
# 8832 - Fix memory leak in datagrid

**Explain the _details_ for making this change. What existing problem does the pull request solve?**
On updateRow function:
Existing DOM elements were not cleaned up before replacing with new ones.
Fixes:
- Detach eventListeners.
- Remove all attributes.
- Remove all attached data.

setTimeout functions have closure errors.
- Fix: Now passing the external variables as function parameter.

eventListener functions have closure errors.
- set the parent function of eventListener's function to null during destroy event of the datagrid.


calculateTextWidth is calculating the entire list cells.
- Fix: Just calculate what is visible in the screen.

**Related github/jira issue (required)**:

Closes https://github.com/infor-design/enterprise/issues/8832

**Steps necessary to review your pull request (required)**:
- test scenarios:
In M3 Angular application:
1. Open 2 programs then scroll down 10x
2. Scroll 10x more 
3. Check chrome task manager
    Expected: No memory leak
    Actual: Memory heap usage has reached 3gb
![memleak](https://github.com/infor-design/enterprise/assets/20462137/e49efec7-b510-4da6-b4d8-35ca04c68b3a)
![memleakTM](https://github.com/infor-design/enterprise/assets/20462137/751c8590-c8e7-4757-8d4c-242b0ec821c6)
https://github.com/infor-design/enterprise/assets/20462137/ec159218-9182-4b1c-8694-652f2777daea
original recording: file://PHMANHBANZON01/Users/hbanzon/OneDrive%20-%20Infor/For%20Share/memleak.webm

**Included in this Pull Request**:
